<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Recommended Work Flow • ewastools</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Recommended Work Flow">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ewastools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.8.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/exemplary_ewas.html">Recommended Work Flow</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Recommended Work Flow</h1>
                        <h4 data-toc-skip class="author">Jonathan
Heiss</h4>
            
            <h4 data-toc-skip class="date">2025-06-20</h4>
      

      <div class="d-none name"><code>exemplary_ewas.Rmd</code></div>
    </div>

    
    
<style type="text/css">
body{ font-size: 12pt; text-align: justify; }
img{ border: 0px;}
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
</style>
<p>This vignette exemplifies how to use the <code>ewastools</code>
package to clean and pre-process DNA methylation data. After loading the
required packages, analysis would start with gathering the phenotype
data. In this example using a public dataset from the <a href="https://www.ncbi.nlm.nih.gov/geo/" class="external-link">Gene Expression Omnibus</a>
repository, the phenotype data is stored in a file named
<code>pheno.csv</code></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pheno</span> <span class="op">=</span> <span class="fu">fread</span><span class="op">(</span><span class="st">"pheno.csv"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pheno</span><span class="op">)</span></span>
<span><span class="co">##           gsm    sex smoker</span></span>
<span><span class="co">##        &lt;char&gt; &lt;char&gt; &lt;char&gt;</span></span>
<span><span class="co">## 1: GSM2260480      m smoker</span></span>
<span><span class="co">## 2: GSM2260482      m smoker</span></span>
<span><span class="co">## 3: GSM2260485      m smoker</span></span>
<span><span class="co">## 4: GSM2260486      m smoker</span></span>
<span><span class="co">## 5: GSM2260487      f smoker</span></span>
<span><span class="co">## 6: GSM2260488      m smoker</span></span></code></pre></div>
<p><code>pheno</code> contains a column <code>gsm</code>, which in this
case represents also the prefix of the .idat file names. Usually,
however, the file names are a combination of the Sentrix barcode,
Sentrix position and color channel and will look something like this
<code>200379120004_R01C01_Red.idat</code> for the .idat containing the
red color channel, and analogously
<code>200379120004_R01C01_Grn.idat</code> for the .idat containing the
green color channel. <code>read_idats</code> can be used to import
methylation data. It’s first argument is a character vector containing
the absolute or relative file paths and names but without the color
channel and file extension,
e.g. <code>C:/folder/subfolder/200379120004_R01C01</code>. Both red and
green .idat files of a particular sample need to be in the same
folder.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meth</span> <span class="op">=</span> <span class="fu"><a href="../reference/read_idats.html">read_idats</a></span><span class="op">(</span><span class="va">pheno</span><span class="op">$</span><span class="va">gsm</span>,quiet<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="co"># `quiet=TRUE` supresses the progress bar</span></span>
<span><span class="co">## [1] 622399</span></span></code></pre></div>
<p>The entire pre-processing, including filtering by detection p-values,
dye-bias correction and conversion into beta-values, can be done in one
line …</p>
<pre><code><span><span class="va">beta</span> <span class="op">=</span> <span class="va">meth</span> <span class="op">%&gt;%</span> <span class="va">detectionP</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/detectionP.html">mask</a></span><span class="op">(</span><span class="fl">0.01</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="va">correct_dye_bias</span> <span class="op">%&gt;%</span> <span class="va">dont_normalize</span></span></code></pre>
<p>… but we will break it up in order to describe the various steps.</p>
<p>The first step should be to filter out unreliable data points which
result from low fluorescence intensities. These can be the result of
insufficiently amplified DNA. Filtering is done using so-called
<strong>detection p-values</strong>, calculated from comparing
fluorescence intensities to a noise distribution. Probes below a chosen
significance threshold are deemed detected, otherwise undetected. The
conventional way of calculating these p-values, as implemented in the
GenomeStudio software, lets many unreliable data points pass,
demonstrated by the fact that many probes targeting the Y chromosome are
classified as detected. <code>ewastools</code> implements an improved
estimation of noise levels that improves accuracy.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meth</span> <span class="op">%&lt;&gt;%</span> <span class="va">detectionP</span></span>
<span><span class="va">P.new</span> <span class="op">=</span> <span class="va">meth</span><span class="op">$</span><span class="va">detP</span></span></code></pre></div>
<p>For easy comparison a function <code><a href="../reference/detectionP.html">detectionP.neg()</a></code> is
provided, which estimates background the conventional way.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">P.neg</span> <span class="op">=</span> <span class="va">meth</span> <span class="op">%&gt;%</span> <span class="va">detectionP.neg</span> <span class="op">%$%</span> <span class="va">detP</span></span></code></pre></div>
<p>We can see the improved accuracy by counting the number of Y
chromosome probes that are called detected for a male and a female
samples.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chrY</span> <span class="op">=</span> <span class="va">meth</span><span class="op">$</span><span class="va">manifest</span><span class="op">[</span><span class="va">chr</span><span class="op">==</span><span class="st">'chrY'</span>,<span class="va">index</span><span class="op">]</span></span>
<span></span>
<span>  <span class="va">male</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">pheno</span><span class="op">$</span><span class="va">sex</span><span class="op">==</span><span class="st">"m"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">female</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">pheno</span><span class="op">$</span><span class="va">sex</span><span class="op">==</span><span class="st">"f"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span></span>
<span><span class="va">P.neg</span> <span class="op">=</span> <span class="va">P.neg</span><span class="op">[</span><span class="va">chrY</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">male</span>,<span class="va">female</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">P.new</span> <span class="op">=</span> <span class="va">P.new</span><span class="op">[</span><span class="va">chrY</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">male</span>,<span class="va">female</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">P.neg</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">P.neg</span><span class="op">&lt;</span><span class="fl">0.01</span>,na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">P.new</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">P.new</span><span class="op">&lt;</span><span class="fl">0.01</span>,na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">P.neg</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"male"</span>,<span class="st">"female"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">P.new</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"male"</span>,<span class="st">"female"</span><span class="op">)</span></span></code></pre></div>
<p>Using the conventional detection p-value, for the female sample 402 Y
chromosome probes are called detected, a number close to all 416 Y
chromosome probes as for the male sample.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">P.neg</span></span>
<span><span class="co">##   male female </span></span>
<span><span class="co">##    416    402</span></span></code></pre></div>
<p>Using the improved method gives a much more accurate result with all
416 Y chromosome probes classified as detected for the male sample, but
only 402 probes classified as detected for the female sample. More
information can be found in <a href="https://doi.org/10.1186/s13148-019-0615-3" class="external-link">Heiss and Just,
2019</a>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">P.new</span></span>
<span><span class="co">##   male female </span></span>
<span><span class="co">##    416     73</span></span></code></pre></div>
<p>We used a significance threshold of 0.01 above. Moving forward,
probes above this threshold should be masked, i.e. set to missing.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">beta</span> <span class="op">=</span> <span class="va">meth</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/detectionP.html">mask</a></span><span class="op">(</span><span class="fl">0.01</span><span class="op">)</span></span></code></pre></div>
<p>Infinium BeadChips use two fluorescent dyes that are linked to the
nucleotides used in the the single-base extension step. A and T
nucleotides use are linked with a red dye (the red color channel), G and
C nucleotides are linked with a green dye (green color channel).
Uncorrected data usually feature higher intensities in the red color
channel, the so-called dye bias. For probes of Infinium type II design,
which use separate color channels to measure the methylated and
unmethylated signal, this results in a shifted distribution of
beta-values. (Probes of Infinium design type I are not affected, as they
measure both signals in the same color channel.) Dye-bias correction
normalizes the red and green color channel. <code>ewastools</code>
provides an improved version of RELIC (<a href="https://doi.org/10.1186/s12864-016-3426-3" class="external-link">Xu et al., 2017</a>)
using robust Theil-Sen estimators.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">beta</span> <span class="op">%&lt;&gt;%</span> <span class="va">correct_dye_bias</span></span></code></pre></div>
<p>The final step is the conversion of intensities to beta-values. While
<code>ewastools</code> implements the LOESS normalization <a href="https://doi.org/10.3389/fgene.2015.00092" class="external-link">(Heiss and Brenner,
2015)</a>, I advise against normalization as it does little to protect
against batch effects but can result in the removal of genuine
biological signal. Instead I recommended to adjust for relevant
technical covariates in regression models later.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">beta</span> <span class="op">%&lt;&gt;%</span> <span class="va">dont_normalize</span></span></code></pre></div>
<p>Before beginning with the actual epigenome-wide association study, it
is advised to check a dataset for problematic samples.</p>
<div class="section level2">
<h2 id="quality-checks">Quality checks<a class="anchor" aria-label="anchor" href="#quality-checks"></a>
</h2>
<p><strong>Control metrics</strong></p>
<p>The first quality check evaluates 17 control metrics which are
describe in the <a href="https://support.illumina.com/content/dam/illumina-support/documents/documentation/chemistry_documentation/infinium_assays/infinium_hd_methylation/beadarray-controls-reporter-user-guide-1000000004009-00.pdf" class="external-link">BeadArray
Controls Reporter Software Guide</a> from Illumina. Exemplary, the
“Bisulfite Conversion II” metric is plotted below. Three samples fall
below the Illumina-recommended cut-off of 1. Input for
<code><a href="../reference/control_metrics.html">control_metrics()</a></code> is the output of
<code><a href="../reference/read_idats.html">read_idats()</a></code>, e.g. the object holding raw or
dye-bias-corrected intensities.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctrls</span> <span class="op">=</span> <span class="fu"><a href="../reference/control_metrics.html">control_metrics</a></span><span class="op">(</span><span class="va">meth</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/stripchart.html" class="external-link">stripchart</a></span><span class="op">(</span><span class="va">ctrls</span><span class="op">$</span><span class="va">`Bisulfite Conversion II`</span>,method<span class="op">=</span><span class="st">"jitter"</span>,pch<span class="op">=</span><span class="fl">4</span>,xlab<span class="op">=</span><span class="st">'Bisulfite Conversion II'</span>,xlim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>v<span class="op">=</span><span class="fl">1</span>,col<span class="op">=</span><span class="fl">2</span>,lty<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="exemplary_ewas_files/figure-html/unnamed-chunk-12-1.png" width="576" style="display: block; margin: auto;"></p>
<p>A logical vector of passed/failed is returned by
<code><a href="../reference/control_metrics.html">sample_failure()</a></code> which compares all 21 metrics against the
thresholds recommended by Illumina. In this case several samples
fail..</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pheno</span><span class="op">$</span><span class="va">failed</span> <span class="op">=</span> <span class="fu"><a href="../reference/control_metrics.html">sample_failure</a></span><span class="op">(</span><span class="va">ctrls</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">pheno</span><span class="op">$</span><span class="va">failed</span><span class="op">)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## FALSE  TRUE </span></span>
<span><span class="co">##    26     9</span></span></code></pre></div>
<hr>
<p><strong>Sex mismatches</strong></p>
<p>The sex of the sample donor can reliable be inferred from the
methylation data. This functionality is implemented by the combination
of <code><a href="../reference/check_sex.html">check_sex()</a></code> and <code><a href="../reference/check_sex.html">predict_sex()</a></code>.
<code><a href="../reference/check_sex.html">check_sex()</a></code> computes the normalized average total
fluorescence intensities of the probes targeting the X and Y chromosome.
<code><a href="../reference/check_sex.html">predict_sex()</a></code> uses the output of <code><a href="../reference/check_sex.html">check_sex()</a></code>
and recorded sex in order to identify mislabeled samples. The function
<code><a href="../reference/check_sex.html">check_sex()</a></code> should be applied to dye-bias corrected
intensities.</p>
<p><em>Important:</em> This test should be performed using dye-bias
corrected intensities <strong>before</strong> masking undetected probes,
as this step would mask many of the Y chromosome probes used here.</p>
<p>Plotted below are the normalized average total fluorescence
intensities of X and Y chromosome probes.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pheno</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"X"</span>,<span class="st">"Y"</span><span class="op">)</span> <span class="op">:=</span> <span class="fu"><a href="../reference/check_sex.html">check_sex</a></span><span class="op">(</span><span class="va">meth</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">pheno</span><span class="op">[</span>,<span class="va">predicted_sex</span><span class="op">:=</span><span class="fu"><a href="../reference/check_sex.html">predict_sex</a></span><span class="op">(</span><span class="va">X</span>,<span class="va">Y</span>,<span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">sex</span><span class="op">==</span><span class="st">"m"</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">sex</span><span class="op">==</span><span class="st">"f"</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">tmp</span> <span class="op">=</span> <span class="va">pheno</span><span class="op">[</span><span class="va">predicted_sex</span><span class="op">==</span><span class="va">sex</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">Y</span> <span class="op">~</span> <span class="va">X</span>,data<span class="op">=</span><span class="va">tmp</span>,pch<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">tmp</span><span class="op">$</span><span class="va">sex</span><span class="op">==</span><span class="st">"f"</span>,<span class="fl">1</span>,<span class="fl">4</span><span class="op">)</span>,asp<span class="op">=</span><span class="fl">1</span>,xlab<span class="op">=</span><span class="st">"Normalized X chromosome intensities"</span>,ylab<span class="op">=</span><span class="st">"Normalized Y chromosome intensities"</span><span class="op">)</span></span>
<span><span class="va">tmp</span> <span class="op">=</span> <span class="va">pheno</span><span class="op">[</span><span class="va">predicted_sex</span><span class="op">!=</span><span class="va">sex</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/points.html" class="external-link">points</a></span><span class="op">(</span><span class="va">Y</span> <span class="op">~</span> <span class="va">X</span>,data<span class="op">=</span><span class="va">tmp</span>,pch<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">tmp</span><span class="op">$</span><span class="va">sex</span><span class="op">==</span><span class="st">"f"</span>,<span class="fl">1</span>,<span class="fl">4</span><span class="op">)</span>,col<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"topright"</span>,pch<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">4</span><span class="op">)</span>,legend<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"female"</span>,<span class="st">"male"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="exemplary_ewas_files/figure-html/unnamed-chunk-14-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Samples form two cluster of males (top left) and females (bottom
left). The one mislabeled sample here (in red) can easily be identified
and should be flagged.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pheno</span><span class="op">[</span><span class="va">sex</span><span class="op">!=</span><span class="va">predicted_sex</span>,<span class="va">exclude</span><span class="op">:=</span><span class="cn">TRUE</span><span class="op">]</span> <span class="co"># flag sample</span></span>
<span><span class="va">pheno</span><span class="op">[</span><span class="va">sex</span><span class="op">!=</span><span class="va">predicted_sex</span>,<span class="fu">.</span><span class="op">(</span><span class="va">gsm</span>,<span class="va">sex</span>,<span class="va">predicted_sex</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">##           gsm    sex predicted_sex</span></span>
<span><span class="co">##        &lt;char&gt; &lt;char&gt;        &lt;fctr&gt;</span></span>
<span><span class="co">## 1: GSM2260573      f             m</span></span></code></pre></div>
<p>Another sample falls outside the two clusters.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pheno</span><span class="op">[</span><span class="va">X</span> <span class="op">%between%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.85</span>,<span class="fl">0.95</span><span class="op">)</span> <span class="op">&amp;</span> <span class="va">Y</span> <span class="op">%between%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.65</span>,<span class="fl">0.75</span><span class="op">)</span>,<span class="fu">.</span><span class="op">(</span><span class="va">gsm</span>,<span class="va">X</span>,<span class="va">Y</span>,<span class="va">sex</span>,<span class="va">predicted_sex</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">##           gsm         X         Y    sex predicted_sex</span></span>
<span><span class="co">##        &lt;char&gt;     &lt;num&gt;     &lt;num&gt; &lt;char&gt;        &lt;fctr&gt;</span></span>
<span><span class="co">## 1: GSM2260653 0.9266193 0.7212944      m             m</span></span></code></pre></div>
<p>There are several possible explanations for samples not clustering
with males or females, for example chromosome abnormalities. Or sample
contamination. The latter theory can be tested in the next quality
check.</p>
<hr>
<p><strong>Genotype calling and outliers</strong></p>
<p>For the next check we first need the row indexes of the SNP probes in
<code>beta</code>. <code>meth</code>, the output of
<code><a href="../reference/read_idats.html">read_idats()</a></code>, contains a data.table object with the
relevant information.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meth</span><span class="op">$</span><span class="va">manifest</span></span>
<span><span class="co">## Index: &lt;channel&gt;</span></span>
<span><span class="co">##            ilmn_id   probe_id addressU addressM probe_design next_base channel</span></span>
<span><span class="co">##             &lt;char&gt;     &lt;char&gt;    &lt;int&gt;    &lt;int&gt;       &lt;fctr&gt;    &lt;fctr&gt;  &lt;fctr&gt;</span></span>
<span><span class="co">##      1: cg00035864 cg00035864 31729416       NA           II      &lt;NA&gt;    Both</span></span>
<span><span class="co">##      2: cg00050873 cg00050873 32735311 31717405            I         A     Red</span></span>
<span><span class="co">##      3: cg00061679 cg00061679 28780415       NA           II      &lt;NA&gt;    Both</span></span>
<span><span class="co">##      4: cg00063477 cg00063477 16712347       NA           II      &lt;NA&gt;    Both</span></span>
<span><span class="co">##      5: cg00121626 cg00121626 19779393       NA           II      &lt;NA&gt;    Both</span></span>
<span><span class="co">##     ---                                                                       </span></span>
<span><span class="co">## 485573:  rs7746156  rs7746156 33622366       NA           II      &lt;NA&gt;    Both</span></span>
<span><span class="co">## 485574:  rs1945975  rs1945975 23614475       NA           II      &lt;NA&gt;    Both</span></span>
<span><span class="co">## 485575:   rs966367   rs966367 16795360       NA           II      &lt;NA&gt;    Both</span></span>
<span><span class="co">## 485576:   rs877309   rs877309 54760445       NA           II      &lt;NA&gt;    Both</span></span>
<span><span class="co">## 485577:  rs4331560  rs4331560 10654345       NA           II      &lt;NA&gt;    Both</span></span>
<span><span class="co">##            chr  mapinfo probe_rep probe_type  index  OOBi     Ui     Mi</span></span>
<span><span class="co">##         &lt;fctr&gt;    &lt;int&gt;     &lt;int&gt;     &lt;fctr&gt;  &lt;int&gt; &lt;int&gt;  &lt;int&gt;  &lt;int&gt;</span></span>
<span><span class="co">##      1:   chrY  8553009         1         cg      1    NA 210047 210047</span></span>
<span><span class="co">##      2:   chrY  9363356         1         cg      2     1 219718 209461</span></span>
<span><span class="co">##      3:   chrY 25314171         1         cg      3    NA 182665 182665</span></span>
<span><span class="co">##      4:   chrY 22741795         1         cg      4    NA  62991  62991</span></span>
<span><span class="co">##      5:   chrY 21664296         1         cg      5    NA  95405  95405</span></span>
<span><span class="co">##     ---                                                                </span></span>
<span><span class="co">## 485573:   &lt;NA&gt;       NA         1         rs 485573    NA 224103 224103</span></span>
<span><span class="co">## 485574:   &lt;NA&gt;       NA         1         rs 485574    NA 126234 126234</span></span>
<span><span class="co">## 485575:   &lt;NA&gt;       NA         1         rs 485575    NA  66829  66829</span></span>
<span><span class="co">## 485576:   &lt;NA&gt;       NA         1         rs 485576    NA 431915 431915</span></span>
<span><span class="co">## 485577:   &lt;NA&gt;       NA         1         rs 485577    NA   2441   2441</span></span></code></pre></div>
<p>SNP probes are labelled <code>"rs"</code>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snps</span> <span class="op">=</span> <span class="va">meth</span><span class="op">$</span><span class="va">manifest</span><span class="op">[</span><span class="va">probe_type</span><span class="op">==</span><span class="st">"rs"</span>,<span class="va">index</span><span class="op">]</span></span>
<span><span class="va">snps</span> <span class="op">=</span> <span class="va">beta</span><span class="op">[</span><span class="va">snps</span>,<span class="op">]</span></span></code></pre></div>
<p>These SNPs are then used as input for <code><a href="../reference/call_genotypes.html">call_genotypes()</a></code>.
This function estimates the parameters of a mixture model consisting of
three Beta distributions representing the heterozygous and the two
homozygous genotypes, and a fourth component, a uniform distribution,
representing outliers. The functions returns posterior probabilities
used for soft classification. When setting the argument
<code>learn=FALSE</code>, a pre-specified mixture model is used. In this
case, we use the pre-specified model as the dataset is quite small and
maximum likelihood estimation might be unstable.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genotypes</span> <span class="op">=</span> <span class="fu"><a href="../reference/call_genotypes.html">call_genotypes</a></span><span class="op">(</span><span class="va">snps</span>,learn<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><code><a href="../reference/call_genotypes.html">snp_outliers()</a></code> returns the average log odds of belonging
to the outlier component across all SNP probes. I recommend to flag
samples with a score greater than -4 for exclusion.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pheno</span><span class="op">$</span><span class="va">outlier</span> <span class="op">=</span> <span class="fu"><a href="../reference/call_genotypes.html">snp_outliers</a></span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span></span>
<span><span class="va">pheno</span><span class="op">[</span><span class="va">outlier</span> <span class="op">&gt;</span> <span class="op">-</span><span class="fl">4</span>,<span class="fu">.</span><span class="op">(</span><span class="va">gsm</span>,<span class="va">X</span>,<span class="va">Y</span>,<span class="va">outlier</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">##           gsm         X         Y    outlier</span></span>
<span><span class="co">##        &lt;char&gt;     &lt;num&gt;     &lt;num&gt;      &lt;num&gt;</span></span>
<span><span class="co">## 1: GSM2260653 0.9266193 0.7212944 -0.8460863</span></span>
<span><span class="va">pheno</span><span class="op">[</span><span class="va">outlier</span> <span class="op">&gt;</span> <span class="op">-</span><span class="fl">4</span>,<span class="va">exclude</span><span class="op">:=</span><span class="cn">TRUE</span><span class="op">]</span> <span class="co"># flag sample</span></span></code></pre></div>
<p>The one sample failing this check is the same sample that did not
belong to either the male or female cluster in the plot above. This is
strong evidence that this sample is indeed contaminated. While SNP
outliers can also result from poorly performing assays, the sample
passed the first quality check looking at the control metrics, therefore
rendering this possibility unlikely. Another cause for a high outlier
score is sample degradation (e.g., FFPE samples).</p>
<p>Other useful functions to be mentioned are
<code><a href="../reference/check_snp_agreement.html">check_snp_agreement()</a></code> and
<code><a href="../reference/check_snp_agreement.html">enumerate_sample_donors()</a></code>. The former checks whether the
genotypes of samples supposed to come from the same donor (or from
monozygotic twins) do in fact agree; the latter returns unique IDs for
unique genotypes and can, for example, be used to find technical
replicates in public datasets.</p>
<p>[Note] <em>When calling <code><a href="../reference/check_snp_agreement.html">check_snp_agreement()</a></code> I
recommend to run the function on all samples with and outlier metric
below -2, i.e., greater than the cut-off of -4 used to exclude
contaminated samples, but still small enough to guarantee that the SNP
calls are accurate.</em></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pheno</span><span class="op">$</span><span class="va">donor_id</span> <span class="op">=</span> <span class="fu"><a href="../reference/check_snp_agreement.html">enumerate_sample_donors</a></span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># List duplicates</span></span>
<span><span class="va">pheno</span><span class="op">[</span>,<span class="va">n</span><span class="op">:=</span><span class="va">.N</span>,by<span class="op">=</span><span class="va">donor_id</span><span class="op">]</span></span>
<span><span class="va">pheno</span><span class="op">[</span><span class="va">n</span><span class="op">&gt;</span><span class="fl">1</span>,<span class="fu">.</span><span class="op">(</span><span class="va">gsm</span>,<span class="va">donor_id</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">##           gsm donor_id</span></span>
<span><span class="co">##        &lt;char&gt;    &lt;num&gt;</span></span>
<span><span class="co">## 1: GSM2260485        3</span></span>
<span><span class="co">## 2: GSM2260543        3</span></span>
<span><span class="va">pheno</span><span class="op">[</span><span class="va">gsm</span><span class="op">==</span><span class="st">"GSM2260543"</span>,<span class="va">exclude</span><span class="op">:=</span><span class="cn">TRUE</span><span class="op">]</span> <span class="co"># drop duplicate</span></span></code></pre></div>
<p>Here samples GSM2260485 and GSM2260543 come from the same donor.</p>
<hr>
<p><strong>PCA</strong></p>
<p>Principal component analysis is a popular feature reduction method:
it projects high-dimensional data into a lower-dimensional
representation while trying to retain as much variability as possible.
This is especially useful when either individual features are highly
correlated and it is therefore reasonable to summarize them, or when
(sometimes subtle) traces of background effects can be found across of
large number of features.</p>
<p>We will drop the X and Y chromosome as we would like to find
important drivers of methylation beyond sex.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">982278</span><span class="op">)</span></span>
<span><span class="va">chrXY</span> <span class="op">=</span> <span class="va">meth</span><span class="op">$</span><span class="va">manifest</span><span class="op">[</span><span class="va">chr</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chrX"</span>,<span class="st">"chrY"</span><span class="op">)</span> <span class="op">&amp;</span> <span class="va">probe_type</span> <span class="op">!=</span> <span class="st">"rs"</span>, <span class="va">index</span><span class="op">]</span></span>
<span><span class="va">pcs</span> <span class="op">=</span> <span class="va">beta</span><span class="op">[</span><span class="op">-</span><span class="va">chrXY</span>, <span class="op">]</span></span>
<span><span class="va">pcs</span> <span class="op">=</span> <span class="va">pcs</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">pcs</span><span class="op">)</span></span>
<span><span class="va">pcs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/na.fail.html" class="external-link">na.omit</a></span><span class="op">(</span><span class="va">pcs</span><span class="op">)</span></span>
<span><span class="va">pcs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">pcs</span><span class="op">)</span></span>
<span><span class="va">pcs</span> <span class="op">=</span> <span class="fu">trlan.svd</span><span class="op">(</span><span class="va">pcs</span>,neig<span class="op">=</span><span class="fl">2</span><span class="op">)</span> <span class="co"># compute the first two principal components</span></span>
<span><span class="va">pcs</span> <span class="op">=</span> <span class="va">pcs</span><span class="op">$</span><span class="va">u</span></span>
<span></span>
<span><span class="va">pheno</span><span class="op">$</span><span class="va">pc1</span> <span class="op">=</span> <span class="va">pcs</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">pheno</span><span class="op">$</span><span class="va">pc2</span> <span class="op">=</span> <span class="va">pcs</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">pc2</span> <span class="op">~</span> <span class="va">pc1</span>,pch<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">sex</span><span class="op">==</span><span class="st">"f"</span>,<span class="fl">1</span>,<span class="fl">4</span><span class="op">)</span>,<span class="va">pheno</span>,asp<span class="op">=</span><span class="fl">1</span>,xlab<span class="op">=</span><span class="st">"PC #1"</span>,ylab<span class="op">=</span><span class="st">"PC #2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"topright"</span>,pch<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">4</span><span class="op">)</span>,legend<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"female"</span>,<span class="st">"male"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="exemplary_ewas_files/figure-html/unnamed-chunk-23-1.png" width="576" style="display: block; margin: auto;"></p>
<p>There is one clear outlier.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pheno</span><span class="op">[</span><span class="va">pc1</span><span class="op">&lt;</span> <span class="op">-</span><span class="fl">0.8</span>,<span class="va">exclude</span><span class="op">:=</span><span class="cn">TRUE</span><span class="op">]</span></span>
<span><span class="va">pheno</span><span class="op">[</span><span class="va">pc1</span><span class="op">&lt;</span> <span class="op">-</span><span class="fl">0.8</span>,<span class="fu">.</span><span class="op">(</span><span class="va">gsm</span>,<span class="va">pc1</span>,<span class="va">pc2</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">##           gsm        pc1        pc2</span></span>
<span><span class="co">##        &lt;char&gt;      &lt;num&gt;      &lt;num&gt;</span></span>
<span><span class="co">## 1: GSM2219539 -0.9300883 -0.1484162</span></span></code></pre></div>
<p>GSM2219539 is actually a lung tissue sample from another GEO dataset
(included here for educational purposes). It dominates the first
principal component and should be excluded as it otherwise could
drastically change the results of downstream analyses.</p>
<p>PCA may be applied iteratively. After excluding samples that manifest
as outliers, repeating PCA can give very different principal
components.</p>
<hr>
<p><strong>Leukocyte composition</strong></p>
<p>This quality check will only apply in case of blood samples (blood
is, however, one of the most commonly studied tissues). The function
<code><a href="../reference/cell_composition.html">estimateLC()</a></code> implements the <a href="https://doi.org/10.1186/1471-2105-13-86" class="external-link">Houseman method</a> to
predict the leukocyte composition. The user has the choice between
various sets of model parameters trained on various reference datasets
(see <code><a href="../reference/cell_composition.html">?estimateLC</a></code> for a list of options). The function
operates on the matrix of beta-values.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">LC</span> <span class="op">=</span> <span class="fu"><a href="../reference/cell_composition.html">estimateLC</a></span><span class="op">(</span><span class="va">beta</span>,ref<span class="op">=</span><span class="st">"HRS"</span><span class="op">)</span></span>
<span><span class="va">pheno</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">pheno</span>,<span class="va">LC</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">LC</span><span class="op">)</span>,<span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">##       NE    EO    BA    MO EM_CT CM_CT  E_CT  N_CT EM_HT CM_HT  E_HT  N_HT</span></span>
<span><span class="co">##    &lt;num&gt; &lt;num&gt; &lt;num&gt; &lt;num&gt; &lt;num&gt; &lt;num&gt; &lt;num&gt; &lt;num&gt; &lt;num&gt; &lt;num&gt; &lt;num&gt; &lt;num&gt;</span></span>
<span><span class="co">## 1: 0.877 0.011 0.000 0.049 0.016 0.006 0.000 0.018     0 0.025 0.000 0.000</span></span>
<span><span class="co">## 2: 0.807 0.007 0.016 0.068 0.002 0.001 0.030 0.017     0 0.019 0.000 0.045</span></span>
<span><span class="co">## 3: 0.888 0.000 0.003 0.033 0.009 0.000 0.000 0.030     0 0.010 0.000 0.000</span></span>
<span><span class="co">## 4: 0.785 0.036 0.013 0.075 0.029 0.010 0.000 0.017     0 0.045 0.000 0.000</span></span>
<span><span class="co">## 5: 0.752 0.024 0.004 0.078 0.003 0.001 0.010 0.026     0 0.042 0.000 0.060</span></span>
<span><span class="co">## 6: 0.783 0.014 0.027 0.050 0.032 0.004 0.006 0.011     0 0.034 0.002 0.032</span></span>
<span><span class="co">##        B    DC    NK</span></span>
<span><span class="co">##    &lt;num&gt; &lt;num&gt; &lt;num&gt;</span></span>
<span><span class="co">## 1: 0.007 0.000 0.000</span></span>
<span><span class="co">## 2: 0.003 0.000 0.007</span></span>
<span><span class="co">## 3: 0.011 0.012 0.020</span></span>
<span><span class="co">## 4: 0.000 0.000 0.010</span></span>
<span><span class="co">## 5: 0.009 0.000 0.007</span></span>
<span><span class="co">## 6: 0.006 0.000 0.007</span></span></code></pre></div>
<p><code>LC</code> contains estimated proportions for seven cell types
(dependent on the chosen reference dataset).</p>
<p>A second foreign sample from another GEO dataset has been hidden in
the dataset, consisting of a purified fraction of granulocytes. Plotting
<code>NE</code> + <code>EO</code> + <code>BA</code> this sample can
easily be spotted.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">pheno</span><span class="op">[</span>,<span class="va">NE</span><span class="op">+</span><span class="va">EO</span><span class="op">+</span><span class="va">BA</span><span class="op">]</span><span class="op">*</span><span class="fl">100</span>,ylab<span class="op">=</span><span class="st">"Granulocyte fraction (%)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="exemplary_ewas_files/figure-html/unnamed-chunk-26-1.png" width="576" style="display: block; margin: auto;"></p>
<p>It is the third to last sample, GSM1185585.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pheno</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.max</a></span><span class="op">(</span><span class="va">NE</span><span class="op">)</span>, <span class="fu">.</span><span class="op">(</span><span class="va">gsm</span>, <span class="va">NE</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">##           gsm       NE</span></span>
<span><span class="co">##        &lt;char&gt;    &lt;num&gt;</span></span>
<span><span class="co">## 1: GSM1185585 1.009256</span></span>
<span><span class="va">pheno</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.max</a></span><span class="op">(</span><span class="va">NE</span><span class="op">)</span>, <span class="va">exclude</span> <span class="op">:=</span> <span class="cn">TRUE</span><span class="op">]</span></span></code></pre></div>
<p>The lung sample is also prominent, with an estimated proportion of
<code>GR</code> of (not exactly because of numerical issues) zero.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pheno</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.min</a></span><span class="op">(</span><span class="va">NE</span><span class="op">)</span>, <span class="fu">.</span><span class="op">(</span><span class="va">gsm</span>, <span class="va">NE</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">##           gsm        NE</span></span>
<span><span class="co">##        &lt;char&gt;     &lt;num&gt;</span></span>
<span><span class="co">## 1: GSM2219539 0.1763932</span></span>
<span><span class="va">pheno</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.min</a></span><span class="op">(</span><span class="va">NE</span><span class="op">)</span>, <span class="va">exclude</span> <span class="op">:=</span> <span class="cn">TRUE</span><span class="op">]</span></span></code></pre></div>
<p>We drop the problematic samples</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">drop</span>  <span class="op">=</span> <span class="va">pheno</span><span class="op">[</span><span class="va">exclude</span><span class="op">==</span><span class="cn">TRUE</span>,which<span class="op">=</span><span class="cn">TRUE</span><span class="op">]</span></span>
<span><span class="va">pheno</span> <span class="op">=</span> <span class="va">pheno</span><span class="op">[</span> <span class="op">-</span><span class="va">drop</span><span class="op">]</span></span>
<span><span class="va">beta</span>  <span class="op">=</span> <span class="va">beta</span> <span class="op">[</span>,<span class="op">-</span><span class="va">drop</span><span class="op">]</span></span>
<span><span class="va">meth</span> <span class="op">%&lt;&gt;%</span> <span class="fu"><a href="../reference/drop_samples.html">drop_samples</a></span><span class="op">(</span>j<span class="op">=</span><span class="va">drop</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="ewas">EWAS<a class="anchor" aria-label="anchor" href="#ewas"></a>
</h2>
<p>You’ve cleaned and pre-processed the data, now it is time for the
actual EWAS. First it is important to correctly code the variables.
<code>smoker</code> and <code>sex</code> are vectors of type
<code>character</code>, but should be converted to factors.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pheno</span> <span class="op">=</span> <span class="va">pheno</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span></span>
<span>     <span class="va">gsm</span></span>
<span>    ,sex <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">sex</span>,levels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"m"</span>,<span class="st">"f"</span><span class="op">)</span><span class="op">)</span> <span class="co"># first level is the reference</span></span>
<span>    ,smoker <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">smoker</span>,levels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"non-smoker"</span>,<span class="st">"smoker"</span><span class="op">)</span><span class="op">)</span></span>
<span>    ,<span class="va">NE</span>, <span class="va">MO</span>, <span class="va">B</span>, <span class="va">EO</span>, <span class="va">BA</span>, <span class="va">NK</span></span>
<span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>We want test all CpG sites for their association with smoking.
Unfortunately, the phenotype data is very sparse, as it is typcial for
public datasets. Aside from smoking, only sex and the estimated
proportions of the seven cell types will be including in the model. The
following code snippet is optimized for readability not speed.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">f</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">meth</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">m</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html" class="external-link">lm</a></span><span class="op">(</span><span class="va">meth</span> <span class="op">~</span> <span class="fl">1</span> <span class="op">+</span> <span class="va">sex</span> <span class="op">+</span> <span class="va">smoker</span> <span class="op">+</span> <span class="va">NE</span> <span class="op">+</span> <span class="va">MO</span> <span class="op">+</span> <span class="va">B</span> <span class="op">+</span> <span class="va">EO</span> <span class="op">+</span> <span class="va">BA</span> <span class="op">+</span> <span class="va">NK</span>, data<span class="op">=</span><span class="va">pheno</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/coef.html" class="external-link">coef</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="st">"smokersmoker"</span>,<span class="fl">4</span><span class="op">]</span> <span class="co"># extract the p-value for the smoking</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">f</span> <span class="op">=</span> <span class="fu">possibly</span><span class="op">(</span><span class="va">f</span>,otherwise<span class="op">=</span><span class="cn">NA_real_</span><span class="op">)</span> <span class="co"># catch errors</span></span>
<span></span>
<span><span class="va">pvals</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">beta</span>,<span class="fl">1</span>,<span class="va">f</span><span class="op">)</span></span></code></pre></div>
<p>We create a <code>data.table</code> holding the p-values and
information about the probes.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SMK</span> <span class="op">=</span> <span class="fu">data.table</span><span class="op">(</span>probe_id<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">beta</span><span class="op">)</span>,pval<span class="op">=</span><span class="va">pvals</span><span class="op">)</span></span>
<span><span class="va">SMK</span> <span class="op">%&lt;&gt;%</span> <span class="va">na.omit</span></span>
<span><span class="va">SMK</span><span class="op">[</span>,<span class="va">qval</span><span class="op">:=</span><span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">pval</span>,m<span class="op">=</span><span class="st">"fdr"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">SMK</span> <span class="op">=</span> <span class="va">SMK</span><span class="op">[</span><span class="va">qval</span><span class="op">&lt;</span><span class="fl">0.05</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">SMK</span><span class="op">)</span></span>
<span><span class="co">##       probe_id         pval        qval</span></span>
<span><span class="co">##         &lt;char&gt;        &lt;num&gt;       &lt;num&gt;</span></span>
<span><span class="co">##  1: cg06882199 7.348569e-07 0.039618668</span></span>
<span><span class="co">##  2: cg01940273 1.389297e-08 0.002247054</span></span>
<span><span class="co">##  3: cg05951221 2.380479e-07 0.019250971</span></span>
<span><span class="co">##  4: cg21566642 7.874678e-09 0.002247054</span></span>
<span><span class="co">##  5: cg05575921 4.688252e-08 0.005687096</span></span>
<span><span class="co">##  6: cg14580211 2.051527e-07 0.019250971</span></span>
<span><span class="co">##  7: cg21161138 1.162694e-08 0.002247054</span></span>
<span><span class="co">##  8: cg25845688 1.247851e-06 0.049582165</span></span>
<span><span class="co">##  9: cg16571007 9.630933e-07 0.046731311</span></span>
<span><span class="co">## 10: cg02610360 1.161149e-06 0.049582165</span></span>
<span><span class="co">## 11: cg03099988 3.273600e-07 0.019855241</span></span>
<span><span class="co">## 12: cg16497398 1.328401e-06 0.049582165</span></span>
<span><span class="co">## 13: cg03636183 3.261651e-07 0.019855241</span></span></code></pre></div>
<p>Two of the significant CpGs (cg01940273, cg05575921) are known
biomarkers overlapping genes (ALPPL2, AHRR) for which the association
with smoking has been validated in several cohorts <a href="https://dx.doi.org/10.1371%2Fjournal.pone.0063812" class="external-link">(Zeilinger et.
al., 2013)</a>.</p>
</div>
<div class="section level2">
<h2 id="final-comments">Final comments<a class="anchor" aria-label="anchor" href="#final-comments"></a>
</h2>
<p>Depending on the dataset, many other types of quality checks might be
applicable. If you have suggestions or comments regarding
<code>ewastools</code>, please send an <a href="mailto:hhhh5@posteo.de">email</a>, or file an issue or submit a
pull request on GitHub (<a href="https://github.com/hhhh5/ewastools" class="external-link uri">https://github.com/hhhh5/ewastools</a>).</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jonathan Alexander Heiss.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
